src/zope/schema/_bootstrapinterfaces.py

```diff
@@ -76,22 +77,94 @@ class RequiredMissing(ValidationError):
 class WrongType(ValidationError):
     __doc__ = _("""Object is of wrong type.""")
 
+    #: The type or tuple of types that was expected.
+    #:
+    #: .. versionadded:: 4.7.0
+    expected_type = None
 
-class TooBig(ValidationError):
+    def __init__(self, value=None, expected_type=None, name=None, *args):
+        """
+        WrongType(value, expected_type, name)
+
+        .. versionchanged:: 4.7.0
+           Added named arguments tot he constructor and the `expected_type`
+           field.
+        """
+        ValidationError.__init__(self, value, expected_type, name, *args)
+        self.expected_type = expected_type
+        self.value = value
+
+
+class OutOfBounds(ValidationError):
+    """
+    A value was out of the allowed bounds.
+
+    This is the common superclass for `OrderableOutOfBounds` and
+    `LenOutOfBounds`, which in turn are the superclasses for `TooBig`
+    and `TooSmall`, and `TooLong` and `TooShort`, respectively.
+
+    .. versionadded:: 4.7.0
+    """
+
+    #: The value that was exceeded
+    bound = None
+
+    #: A constant for `violation_direction`.
+    TOO_LARGE = object()
+
+    #: A constant for `violation_direction`.
+    TOO_SMALL = object()
```

Comment:
I would like these to have programmer-friendly `__repr__`s.  Maybe something like 

```python
TOO_LARGE = type('TOO_LARGE', (), {})()
```

to avoid overengineering with custom classes etc?

Maybe when Python 2 finally dies we can use `enum`s?

Reply:
Good idea.